llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -2518,7 +2518,12 @@ bool SelectionDAG::expandMultipleResultFPLibCall(
     SDValue StoreValue = ST->getValue();
     unsigned ResNo = StoreValue.getResNo();
     Type *StoreType = StoreValue.getValueType().getTypeForEVT(Ctx);
-    if (CallRetResNo == ResNo || !ST->isSimple() ||
+    // If the pointer value does not come from the IR, it could come from ABI
+    // lowering and may alias with the arguments of the library call if they are
+    // passed via the stack.
```

Comment:
My understanding (possibly wrong) is you have two calls:
```
sincos(x, ptr, ptr)
g(double, double)
```

All the arguments are passed on the stack, and the call to `sincos` is written to be passed the stack locations of the arguments to `g` directly. The issue is the slack locations of `g`'s arguments alias the stack locations of `sincos`'s arguments, so when `sincos` writes to one of the pointers it clobbers its own arguments. 

The current code is already checking the two pointers for sin and cos don't alias, but it does not expect the pointers could alias its other arguments.

Reply:
>  The correct way to handle that case should involve callseq_start/callseq_end. Assuming the libcall isn't nested inside another callseq_start/end, there should be a callseq_end and a callseq_start between the libcall and the store.

It looks like this is the case. Would a potential fix be to check that a `CALLSEQ_START` does not occur in the chain of the store?

Reply:
You shouldn't have to special case CALLSEQ_START. This restriction should be implicitly present in the chain

Reply:
Having thought about the issue with folding stores into `sincos` while expanding it to a library call, the possible solutions I came up with are:

1. Check the stores are not within a `CALLSEQ_START`-`CALLSEQ_END` pair
   * As far as I can tell nobody has done this before -- grepping the codebase (and it's possibly a costly graph search)
2. Disallow stores to pointers that don't exist in the IR
    * Much simpler -- and I think could solve the issue, but it's assuming a store to an IR pointer won't occur within a `CALLSEQ_START` - `CALLSEQ_END` pair
     * Maybe this is a valid assumption (the stores I've seen have always been to special stack slots for the call), but maybe this could be broken.
3. Be very conservative and restrict this to trivial cases (e.g. the case where the store input chains are the entry node)

The issue this is trying to avoid is if the stores that are to be folded into `sincos` are within a  `CALLSEQ_START`-`CALLSEQ_END` pair, then the expansion will result in nested call sequences, which is where the issues come from.